waspc/src/Wasp/Generator/NpmInstall.hs

```diff
@@ -0,0 +1,119 @@
+module Wasp.Generator.NpmInstall
+  ( ensureNpmInstall,
+  )
+where
+
+import Control.Concurrent (Chan, newChan, readChan)
+import Control.Concurrent.Async (concurrently)
+import Control.Monad (when)
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString.Lazy as B
+import StrongPath (Abs, Dir, File', Path', Rel, relfile, (</>))
+import qualified StrongPath as SP
+import System.Directory (doesFileExist, removeFile)
+import System.Exit (ExitCode (..))
+import Wasp.AppSpec (AppSpec)
+import Wasp.Generator.Common (ProjectRootDir)
+import qualified Wasp.Generator.Job as J
+import Wasp.Generator.Job.IO (printPrefixedJobMessage)
+import Wasp.Generator.Monad (GeneratorError (..), GeneratorWarning (..))
+import qualified Wasp.Generator.NpmDependencies as N
+import Wasp.Generator.ServerGenerator as SG
+import qualified Wasp.Generator.ServerGenerator.Setup as ServerSetup
+import Wasp.Generator.WebAppGenerator as WG
+import qualified Wasp.Generator.WebAppGenerator.Setup as WebAppSetup
+
+-- | Run npm install if needed
+
+-- It only does this if the dependencies described in the user wasp file are
+-- different from the dependencies that we just installed. To this end, this
+-- code keeps track of the dependencies installed a file, which it updates
+-- after each install.
```

Comment:
```suggestion
-- It tries to avoid redunant runs of @npm install@ by running it only if the
-- dependencies described in AppSpec are different from the latest installed dependencies.
-- If they are the same as the last time, there is no need to run @npm install@, since it was already run before.
--
-- To this end, this code keeps track of the latest installed dependencies via a metadata file, which it updates after each new install.
--
-- NOTE: There is an important assumption being made here: We assume here, and also in the rest of the functions in this file, that dependencies from AppSpec are in sync with package.json files inside the generated code on the disk. That assumption should hold since generated code was generated based on AppSpec, and this function should be called after the code generation has completed. One way to avoid having this assumption in the future would be to not obtain dependencies from AppSpec, but to instead read the dependencies directly from the generated package.json files. We decided to stick with the current approach for now.
```